docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+plugin. We propose a similar approach to pluggable processors as we already have with
+pluggable connectors. This means that processors would be loaded at startup and added
+to the processor plugin registry. Processors would be addressable using a similar naming
+scheme as connectors (more on
+[referencing connectors](https://conduit.io/docs/connectors/referencing)). Built-in
+processors would be created in a similar manner as built-in connectors, in that they
+would implement the same interface and could essentially be extracted into standalone
+processors (more info on
+[standalone vs built-in connectors](https://conduit.io/docs/connectors/behavior#standalone-vs-built-in-connectors)).
+
+The only difference between pluggable processors and connectors is the proposed plugin
+mechanism. For pluggable processors we are leaning towards using
+[WebAssembly](https://webassembly.org). We are not decided yet which runtime to use, but
+it will most likely be either [wazero](https://github.com/tetratelabs/wazero) or
+[wasmtime-go](https://github.com/bytecodealliance/wasmtime-go). The benefit of wazero is
+the fact that it is written in pure Go and does not require CGO. On the other hand,
+wasmtime-go is using [wasmtime](https://github.com/bytecodealliance/wasmtime) behind the
+scenes, which is a popular runtime written in Rust and might provide better performance.
+
+The final decision on the chosen runtime will be done as part of the implementation.
+
+We will also need to provide an SDK for users who want to implement their own processor.
+Our proposal is to again mimic the structure we use for connectors and create a separate
+module (e.g. `github.com/conduitio/conduit-processor-sdk`) that exposes a simple SDK for
+implementing a processor plugin. This SDK should provide utilities for logging and
+calling exposed host functions. An example of such an SDK is the
+[Redpanda transform SDK](https://github.com/redpanda-data/redpanda/tree/dev/src/transform-sdk/go).
+
+### Conditional execution of processors
+
+Currently the only way to control what record gets passed to the processor is by
+attaching the processor either to a source or a destination connector. It would
+be useful if the user could additionally define a condition that must be fulfilled for
+the record to get passed to a processor.
+
+We already have a similar functionality in the `filterfieldkey` processor which lets you
+filter a record if some condition is met. The idea is to extract this functionality and
+create a generalized way to create conditions on processors. If the condition evaluates
+to `false`, the processor is skipped and the record is passed along to the next node
+without being processed.
+
+This way we can simplify the `filter` processor and, more importantly, give the user
+the control of defining when a processor should be executed.
+
+We propose adding a field `condition` in the processor configuration that allows you to
```

Comment:
And I guess a failure to evaluate the condition is the same as any other failure in the pipeline and counts towards the DLQ quota?

Reply:
I think so, yes.